objax/util/util.py

```diff
@@ -69,6 +69,32 @@ def local_kwargs(kwargs: dict, f: Callable) -> dict:
 map_to_device: Callable[[List[jn.ndarray]], List[ShardedDeviceArray]] = jax.pmap(lambda x: x, axis_name='device')
 
 
+def merge_args_kwargs(args: Iterable, kwargs: dict, f: Callable) -> Tuple[List, dict]:
+    """Properly merge positional and keyword arguments according to signature of the function.
+
+    If some of the keyword arguments is not in the signature of the function, they will be removed.
+    If some argument is both in args and kwargs, then value from kwargs will be used.
+
+    Args:
+        args: positional arguments.
+        kwargs: keyword arguments.
+        f: callable, for which arguments should be fixed.
+
+    Return:
+        args: corrected list of positional arguments.
+        kwargs: corrected dictionary of keyword arguments.
+    """
+    args = list(args)
+    kwargs = local_kwargs(kwargs, f)
+    p = inspect.signature(f).parameters
+    for idx, (k, v) in enumerate(p.items()):
+        if (idx >= len(args)) or (v.kind == inspect.Parameter.VAR_POSITIONAL):
+            break
+        if k in kwargs:
+            args[idx] = kwargs.pop(k)
+    return args, kwargs
```

Comment:
I propose a significant change here since the original implementation was hiding code bugs.

Here's an example:
```python
class M(Module):
    def __call__(self, a, b, c):
        return 0

g = objax.ForceArgs(M(), b=1)
g(1, 2, 3, b=4) # should error but didn't in the original code.
f(1, 2, 3, b=4) # the ForceArgs version should error just like the original function f.
```

There were other failure scenarios. But the gist is that when passing improper/invalid argument, the behavior of a `ForceArgs(f)` should be the one of `f`. Consequently, I made the following suggestion to only override the value of an argument if present without modifying arguments in any other way.

```suggestion
def override_args_kwargs(args: Tuple[Any, ...],
                         kwargs: Dict[str, Any],
                         override:  Dict[str, Any],
                         f: Callable) -> Tuple[List, Dict[str, Any]]:
    """Override positional and keyword arguments according to signature of the function.

    Replace the values in args and kwargs with the corresponding override entries.

    Args:
        args: positional arguments.
        kwargs: keyword arguments.
        override: arguments for which to override the value.
        f: callable, for which arguments should be fixed.

    Return:
        args: updated list of positional arguments.
        kwargs: updated dictionary of keyword arguments.
    """
    args = list(args)
    params = inspect.signature(f).parameters
    override = override.copy()
    for i, (k, v) in enumerate(itertools.islice(params.items(), len(args))):
        if v.kind not in (inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD):
            break
        if k in override:
            args[i] = override.pop(k)
    return args, {**kwargs, **override}
```

Reply:
Done